chore: update api test suite and include it in run test

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

It looks like a lot, but the main thing this PR does is fix up the supertest tests so that the server does not need to be running in the background before we run the tests.

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
refactor: separate building app from starting it

ojeytonwilliams committed:
refactor: index -> server

ojeytonwilliams committed:
test(api): make server test use running server

ojeytonwilliams committed:
chore: run api tests in ci

ojeytonwilliams committed:
fix(api): relax env requirements in testing

ShaunSHamilton commented:
_i want to get in a few more prs, before we activate tests, and they hinder my bugs_

ojeytonwilliams committed:
fix: use longer default secret + disallow in prod

ojeytonwilliams committed:
chore: run api tests before curriculum

That way we get feedback earlier and don't have to worry about the
flakiness of the curriculum tests if we just care about testing code.

ojeytonwilliams committed:
test: increase timeout to let server start

ojeytonwilliams committed:
chore(ci): create mongod service with replica set

ojeytonwilliams committed:
fix(ci): pin mongodb version

ojeytonwilliams committed:
fix: add mongo service where needed

raisedadead commented on .github/workflows/node.js-tests-upcoming.yml:
```diff
@@ -68,6 +68,7 @@ jobs:
       fail-fast: false
       matrix:
         node-version: [18.x]
+        mongodb-version: [6.0]
```
Does this have to be in the matrix? I do not see us testing in multiple versions of MongoDB, and it means we have to update every open PR for the updated check status this entails.

raisedadead replied:
An alternative would be to use an exact match in the task stanza as an environment variable. I do not know your reasoning here, so I'll defer to what you think is better without using the matrix.

ojeytonwilliams replied:
Nope, I just assumed you'd want to! I'll take them out.

ojeytonwilliams committed:
fix: remove mongodb-version from matrix

Co-authored-by:  Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>

ShaunSHamilton commented on api/app.ts:
```diff
@@ -18,38 +25,26 @@ import {
-}).withTypeProvider<TypeBoxTypeProvider>();
-
-export type FastifyInstanceWithTypeProvider = typeof fastify;
+export type FastifyInstanceWithTypeProvider = FastifyInstance<
+  RawServerDefault,
+  RawRequestDefaultExpression,
+  RawReplyDefaultExpression,
+  FastifyBaseLogger,
+  TypeBoxTypeProvider
+>;
```
Just so we are on the same page (out of interest): Why this?

ojeytonwilliams replied:
Limitations of my TypeScript skills, basically.  We want to correctly type the `fastify` instance, but we're using the `TypeBoxTypeProvider`, so we can't just use `FastifyInstance`'s defaults as they include `FastifyTypeProviderDefault`.

Before we could just use `typeof fastify`, but now `fastify` is created in a function and you can't export from functions.

Anyway, if there's a neater way to generate this, I don't know it.

ShaunSHamilton commented on api/server.ts:
```diff
@@ -0,0 +1,33 @@
+};
+
+const start = async () => {
+  const fastify = await build({ logger: envToLogger[NODE_ENV] });
```
We won't tell Mrugesh about this one 🤫

ojeytonwilliams committed:
fix(api): update test script

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ShaunSHamilton reviewed (approved):
LGTM 👍
